### PR TITLE
[docs] Why TestCafe? + How It Works

### LINKED ISSUE #4033: [docs] Describe TestCafe benefits

### PR DESCRIPTION
**Pull request by titerman**
Splits the "How It Works" article in two
Fixes #4033

### CODE REVIEW

File: docs/articles/documentation/how-it-works/README.md [hunk 1]
```diff
@@ -3,151 +3,42 @@ layout: docs
...
+TestCafe’s engine is an invisible automation layer that takes over browsers and the web pages they display.
 
-* **Dependency on third-party tools.** This required you to install, configure, and maintain additional drivers, SDKs and/or libraries in every test environment.
-* **Limited browser support.** The range of available browsers and platforms was restricted to those supported by the driver.
-* **Boilerplate code.** These testing tools functioned with the same browser instance during the entire test run, so the tested web app state (cookies, storages, etc.) persisted from test to test. As a result, end-to-end tests contain a large amount of boilerplate code that clears the state to avoid interference between tests.
-* **Low automation.** Legacy tools were unable to interact directly with the tested page. Although the fact that the user’s test code did not intervene in the app code was an advantage, the testing tool itself could benefit from operating inside the page. This would enable the tool to automate many routine tasks.
+## The Server-Client Architecture
 
-## TestCafe is Different
+TestCafe is a server-side application with client-side access.
```
[the remark this answers is not in the document]
Comment (VasilyStrelyaev) [reply docs/articles/documentation/how-it-works/README.md:12]: What about smth like *TestCafe runs tests in a standalone process but also provides means to access the page DOM in the browser*?

File: docs/articles/documentation/how-it-works/README.md [hunk 2]
```diff
@@ -3,151 +3,42 @@ layout: docs
...
-* **Low automation.** Legacy tools were unable to interact directly with the tested page. Although the fact that the user’s test code did not intervene in the app code was an advantage, the testing tool itself could benefit from operating inside the page. This would enable the tool to automate many routine tasks.
+## The Server-Client Architecture
 
-## TestCafe is Different
+TestCafe is a server-side application with client-side access. 
 
-The idea that defined TestCafe architecture was that you don't really need an external driver to run end-to-end tests in the browser. Instead, the page can run all the test scripts that emulate user actions. This would allow for true cross-browser and cross-platform tests that run on any device that has a modern browser. This vision eventually developed into the following design concepts.
+TestCafe controls browser processes on a system level. This is necessary to launch browsers, manage browser windows, and have full control over test execution.
```
[the remark this answers is not in the document]
Comment (VasilyStrelyaev) [reply docs/articles/documentation/how-it-works/README.md:14]: I'll just suggest a way of thinking: *TestCafe interacts with the browser processes locally (via local/system means) on a test machine*

File: docs/articles/documentation/how-it-works/README.md [hunk 3]
```diff
@@ -3,151 +3,42 @@ layout: docs
...
+## Browser Sandboxing
 
-Script injection enables TestCafe key features like mobile testing, automatic waiting, or user roles for smart authentication out of the box.
+Every TestCafe test begins with an empty slate of a browser. At the end of the run, TestCafe deletes all browser cookies, empties the storage, and reloads the page, thereby preventing undesirable interference with subsequent tests. You don’t need to write any boilerplate code to reset the app state and reverse the changes your tests make.
 
-### Isolated tests
+Tests that run in parallel operate in independent sandboxed environments. This helps prevent server-side collisions.
 
-Each test starts with a clean browser state. Cookies and storages are reset, an empty user profile is used, and the tested page is reloaded. This helps prevent interference between tests without extra boilerplate code.
+## Arbitrary Client-Side Injections
```
[the remark this answers is not in the document]
Comment (VasilyStrelyaev) [reply docs/articles/documentation/how-it-works/README.md:36]: `Custom (User-Defined) Client Code` ?

**Review by VasilyStrelyaev — CHANGES_REQUESTED** [review]

File: docs/articles/documentation/why-testcafe/README.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+### Debug mode
+Pause your tests to examine the app and troubleshoot errors.
+
+## Community-driven
+
+### Free and Open Source
+TestCafe is distributed for free under the MIT license.
+
+### Community-driven
```
Comment (AlexSkorkin) [inline docs/articles/documentation/why-testcafe/README.md:56]: I don't like two equal and repeated headers:
## Community-Driven
### Community-Driven
I would rename the second 'Community-Driven' sub-header something like 'Quick Responses'

File: docs/articles/documentation/why-testcafe/README.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+The TestCafe team is quick to respond to GitHub issues and StackOverflow questions. Community suggestions shape the future of the product.
+
+### Support
+Paying customers gain access to priority support and receive bugfixes before these changes go live.
```
Comment (AlexSkorkin) [inline docs/articles/documentation/why-testcafe/README.md:60]: ```Paying customers gain access to priority support and receive bugfixes before these changes go live.```

Not quite correct.
- Paying customers gain access to two types of official DevExpress support: standard and priority. Priority support is given for the Universal Subscribers only, not for everyone.
- Bugfixes receive all users, including trial users.
- 'before these changes go live' - it's unclear what you mean. I assume you meant 'these changes are officially released'.

File: docs/articles/documentation/how-it-works/README.md [hunk 6]
```diff
@@ -33,12 +31,12 @@ Every TestCafe test begins with an empty slate of a browser. At the end of the r
```
Comment (AlexSkorkin) [inline docs/articles/documentation/how-it-works/README.md:31]: > This is why you can  your website’s URLs as is

You can what?

File: docs/articles/documentation/how-it-works/README.md [hunk 7]
```diff
@@ -33,12 +31,12 @@ Every TestCafe test begins with an empty slate of a browser. At the end of the r
...
 
-## Arbitrary Client-Side Injections
+## Client-Side Scripts
 
-To perform common testing tasks, TestCafe translates server-side calls to its API into client-side code. However, some scenarios require the execution of client-side code that TestCafe cannot automatically generate. That’s why the framework provides several ways to execute arbitrary JavaScript.
+To perform common testing tasks, TestCafe translates server-side calls to its API into client-side code. However, some scenarios require the execution of client-side code that TestCafe cannot automatically generate. That’s why the framework provides several ways to execute user-defined JavaScript.
 
 [Client Scripts](/../guides/advanced-guides/inject-client-scripts.md) inject custom JavaScript files into the page.
-[Client Functions](../guides/basic-guides/obtain-client-side-info.md) evaluate arbitrary JavaScript expressions and pass their return value to the server side.
+[Client Functions](../guides/basic-guides/obtain-client-side-info.md) evaluate user-defined JavaScript expressions and pass their return value to the server side.
```
Comment (AlexSkorkin) [inline docs/articles/documentation/how-it-works/README.md:39]: This information does not make it clear on why and when use Client Functions.

File: docs/articles/documentation/why-testcafe/README.md (hunk 4 above)
Comment (titerman) [reply docs/articles/documentation/why-testcafe/README.md:56]: oops, my bad. Thanks for pointing this out!

File: docs/articles/documentation/how-it-works/README.md (hunk 7 above)
Comment (titerman) [reply docs/articles/documentation/how-it-works/README.md:39]: Neither does the previous line: this list, as I imagined it, was not meant to describe use cases, but rather make the reader aware of the opportunities they have when it comes to client-side injections. Should I change both lines to fit this approach?

File: docs/articles/documentation/how-it-works/README.md [hunk 8]
```diff
@@ -3,151 +3,40 @@ layout: docs
...
+* TestCafe uses [high-level system APIs](https://github.com/DevExpress/testcafe-browser-tools) to launch and manage browsers. This is necessary to achieve full control over test execution.
+* TestCafe tests are full-fledged Node.js scripts. They can launch services and applications, read and write system files, make use of your favourite libraries.
+* At the same time, TestCafe is capable of performing in-browser actions via client-side automation scripts. This is how TestCafe handles asynchronous events, simulates user activity, and executes custom (user-defined) JavaScript.
 
-### Client-server architecture
+Our experience shows that this hybrid approach gives users the best of both worlds. Early in the product’s life, TestCafe ran entirely in the browser. This caused a number of issues. The tests’ capabilities were limited, the application often interfered with page execution, and the sub-optimal logic separation caused browser crashes. Since then, we decoupled the engine, exposed the underlying Node.js API, and implemented a test runner that executes test scripts server-side.
```
Comment (DIRECTcut) [inline docs/articles/documentation/how-it-works/README.md:18]: > Early in the product’s life, TestCafe ran entirely in the browser.

This line got me thinking 'which product?' for a second.

Perhaps we could do it shorter, like this:

> Earlier versions of TestCafe ran entirely in the browser.

File: docs/articles/documentation/how-it-works/README.md [hunk 9]
```diff
@@ -3,151 +3,40 @@ layout: docs
...
+Every TestCafe test begins with an empty slate of a browser. At the end of a run, TestCafe deletes all browser cookies, empties the storage, and reloads the page, thereby preventing undesirable interference with subsequent tests. You don’t need to write any boilerplate code to reset the app state and reverse the changes your tests make.
 
-To provide automation scripts (client-side code) to the browser, TestCafe runs a reverse proxy under the hood. It proxies the tested page on the server and injects the scripts into its code.
+Tests that run in parallel operate in independent sandboxed environments. This helps prevent server-side collisions.
 
-Script injection enables TestCafe key features like mobile testing, automatic waiting, or user roles for smart authentication out of the box.
+## Client-Side Scripts
 
-### Isolated tests
+To perform common testing tasks, TestCafe translates server-side calls to its API into client-side code. However, some scenarios require the execution of client-side code that TestCafe cannot automatically generate. That’s why the framework provides several ways to execute user-defined JavaScript.
```
Comment (DIRECTcut) [inline docs/articles/documentation/how-it-works/README.md:36]: Ray dislikes 'provides' (as far as I know);
I think 'several' is not comprehensive. I'd replace it:

> That’s why the framework has more than one way to execute user-defined JavaScript.

File: docs/articles/documentation/how-it-works/README.md (hunk 7 above)
Comment (AlexSkorkin) [reply docs/articles/documentation/how-it-works/README.md:39]: I would change them. Let's ask others for their opinion.

File: docs/articles/documentation/how-it-works/README.md (hunk 8 above)
Comment (titerman) [reply docs/articles/documentation/how-it-works/README.md:18]: Great suggestion!

Comment (titerman) [reply docs/articles/documentation/how-it-works/README.md:18]: On a second thought, '_earlier versions_' can be interpreted as 'all versions of the product up until now', which would be clearly false. Let me think

File: docs/articles/documentation/how-it-works/README.md (hunk 9 above)
Comment (titerman) [reply docs/articles/documentation/how-it-works/README.md:36]: what about 'offers multiple ways'?

File: docs/articles/documentation/how-it-works/README.md [hunk 10]
```diff
@@ -3,151 +3,40 @@ layout: docs
...
 
-Test code uses compact API that consists of just a few dozen methods that eventually transform to user actions on the page. Since this server-side code runs in Node.js, tests should be written in JavaScript or a language TestCafe can compile to JavaScript (like TypeScript or CoffeeScript).
+The proxying mechanism ensures that the page appears to be hosted at the original URL even to the test code. This is why you can use your website’s actual URL in tests, and pay no mind to the browser’s address bar.
 
-The concise API makes tests easy to write, read, and maintain. Support for popular web languages allows most users to use their existing skills and minimizes the learning curve for new users.
+## Browser Sandboxing
 
-### Page proxying
+Every TestCafe test begins with an empty slate of a browser. At the end of a run, TestCafe deletes all browser cookies, empties the storage, and reloads the page, thereby preventing undesirable interference with subsequent tests. You don’t need to write any boilerplate code to reset the app state and reverse the changes your tests make.
```
Comment (AlexSkorkin) [inline docs/articles/documentation/how-it-works/README.md:30]: ```an empty slate of a browser```

What is `slate`? My dictionary does not give any good translations for this word. Did you mean `state`?

**Comment by dmcgill50** [comment]
Clean Slate reads better.

David

**Review by AlexSkorkin — APPROVED** [review]
Lets try this version...

**Review by helen-dikareva — APPROVED** [review]

**Review by DIRECTcut — APPROVED** [review]

**Comment by titerman** [comment]
@AlexSkorkin tbh, the 'clean slate' sentence is not really necessary. It doesn't contain any information that can't be found in the other two sentences from the paragraph. The phrasing, as it is right now, is quite weird, so I'm just going to remove the sentence altogether. I hope that's fine.

File: docs/articles/documentation/how-it-works/README.md [hunk 11]
```diff
@@ -3,151 +3,40 @@ layout: docs
...
 
-* **Dependency on third-party tools.** This required you to install, configure, and maintain additional drivers, SDKs and/or libraries in every test environment.
-* **Limited browser support.** The range of available browsers and platforms was restricted to those supported by the driver.
-* **Boilerplate code.** These testing tools functioned with the same browser instance during the entire test run, so the tested web app state (cookies, storages, etc.) persisted from test to test. As a result, end-to-end tests contain a large amount of boilerplate code that clears the state to avoid interference between tests.
-* **Low automation.** Legacy tools were unable to interact directly with the tested page. Although the fact that the user’s test code did not intervene in the app code was an advantage, the testing tool itself could benefit from operating inside the page. This would enable the tool to automate many routine tasks.
+## The Server-Client Architecture
 
-## TestCafe is Different
+The server-client architecture of TestCafe makes it capable of executing both system-level and in-browser code.
```
Comment (VasilyStrelyaev) [inline docs/articles/documentation/how-it-works/README.md:12]: Are we sure it's `system-level` code, not `standalone` or smth? I'm not 100% sure what `system-level code` means.

Like:
*makes it capable of executing both standalone and in-browser code.*
*makes it capable of executing code both in a standalone process and in the browser.*

File: docs/articles/documentation/why-testcafe/README.md [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+### Frames and windows
+Make use of multiple browser windows to test out complex user interactions. Switch to and from iframes with ease and stability.
+
+### Concurrent tests
+Run your tests in multiple browsers at once to quickly uncover browser-specific errors.
+
+## Automate what’s possible
+
+### Automated waiting
+Manual timeouts are a thing of the past. Asynchronous from the ground-up, TestCafe knows when to wait and what to wait for before the test can continue.
```
Comment (VasilyStrelyaev) [inline docs/articles/documentation/why-testcafe/README.md:32]: Can we say that manual timeouts are a thing of the past and... implement [t.wait](https://devexpress.github.io/testcafe/documentation/reference/test-api/testcontroller/wait.html) at the same time?

The message is right, but maybe we can say something less unconditional, like *Manual timeouts make your tests unstable.*

**Review by VasilyStrelyaev — APPROVED** [review]
In general, lgtm, just two small issues

File: docs/articles/documentation/why-testcafe/README.md (hunk 12 above)
Comment (titerman) [reply docs/articles/documentation/why-testcafe/README.md:32]: Can we simply remove this sentence? Or do we need to justify the automated waiting feature?

File: docs/articles/documentation/how-it-works/README.md [hunk 13]
```diff
@@ -3,151 +3,40 @@ layout: docs
...
 
-Test code uses compact API that consists of just a few dozen methods that eventually transform to user actions on the page. Since this server-side code runs in Node.js, tests should be written in JavaScript or a language TestCafe can compile to JavaScript (like TypeScript or CoffeeScript).
+The proxying mechanism ensures that the page appears to be hosted at the original URL even to the test code. This is why you can use your website’s actual URL in tests, and pay no mind to the browser’s address bar.
 
-The concise API makes tests easy to write, read, and maintain. Support for popular web languages allows most users to use their existing skills and minimizes the learning curve for new users.
+## Browser Sandboxing
 
-### Page proxying
+At the end of each run, TestCafe deletes all browser cookies, empties the storage, and reloads the page, thereby preventing undesirable interference with subsequent tests. You don’t need to write any boilerplate code to reset the app state and reverse the changes your tests make.
```
Comment (arminal) [inline docs/articles/documentation/how-it-works/README.md:30]:
```suggestion
At the end of each run, TestCafe deletes all browser cookies, empties the storage, and reloads the page, thereby preventing undesirable interference with subsequent tests. You don’t need to write boilerplate code to reset the app state and reverse the changes your tests make.
```

File: docs/articles/documentation/how-it-works/README.md [hunk 14]
```diff
@@ -3,151 +3,40 @@ layout: docs
...
 
-To provide automation scripts (client-side code) to the browser, TestCafe runs a reverse proxy under the hood. It proxies the tested page on the server and injects the scripts into its code.
+Tests that run in parallel operate in independent sandboxed environments. This helps prevent server-side collisions.
 
-Script injection enables TestCafe key features like mobile testing, automatic waiting, or user roles for smart authentication out of the box.
+## Client-Side Scripts
 
-### Isolated tests
+To perform common testing tasks, TestCafe translates server-side calls to its API into client-side code. However, some scenarios require the execution of client-side code that TestCafe cannot automatically generate. That’s why the framework offers multiple ways to execute user-defined JavaScript.
```
Comment (arminal) [inline docs/articles/documentation/how-it-works/README.md:36]: can you replace the word "some" with something specific?

File: docs/articles/documentation/how-it-works/README.md [hunk 15]
```diff
@@ -3,151 +3,40 @@ layout: docs
...
-* reloads the tested page
-
-This saves you from having to write code to reset the app state and take care of the changes your tests make. As a result, your tests are stable, easy to write and read, and free from boilerplate code.
-
-If you run several tests in parallel, each test run obtains an independent server-side context. This prevents server-side collisions as well.
-
-## Conclusion
-
-TestCafe is a self-sufficient end-to-end testing solution that combines unmatched ease-of-use with advanced automation and robust built-in stability mechanisms. TestCafe officially supports all major browsers and platforms and it is designed to run in most modern browsers beyond officially supported.
+However, client-side injections have their limits. Client-side page modifications can disrupt internal TestCafe processes. Client Functions can not return DOM elements. The rule of thumb is to only inject client-side code when you need to pass otherwise unobtainable page data to the server.
```
Comment (arminal) [inline docs/articles/documentation/how-it-works/README.md:42]: cannot

File: docs/articles/documentation/why-testcafe/README.md [hunk 16]
```diff
@@ -0,0 +1,60 @@
+# Why TestCafe?
+
+TestCafe excels where other end-to-end testing solutions fall short.
+
+## One NPM package
+
+TestCafe doesn't require the use of extra tools and frameworks. All you need to get up and running is a recent version of Node.js, a browser — any browser, and the TestCafe NPM package.
```
Comment (arminal) [inline docs/articles/documentation/why-testcafe/README.md:7]: i don't like the casual tone of your writing 
please rewrite

File: docs/articles/documentation/why-testcafe/README.md (hunk 12 above)
Comment (arminal) [inline docs/articles/documentation/why-testcafe/README.md:32]:
```suggestion
Manual timeouts are a thing of the past. Asynchronous from the ground-up, TestCafe knows when to wait and what to wait for before a test can continue.
```

File: docs/articles/documentation/how-it-works/README.md [hunk 17]
```diff
@@ -3,151 +3,40 @@ layout: docs
...
+* TestCafe uses [high-level system APIs](https://github.com/DevExpress/testcafe-browser-tools) to launch and manage browsers. This is necessary to achieve full control over test execution.
+* TestCafe tests are full-fledged Node.js scripts. They can launch services and applications, read and write system files, make use of your favourite libraries.
+* At the same time, TestCafe is capable of performing in-browser actions via client-side automation scripts. This is how TestCafe handles asynchronous events, simulates user activity, and executes custom (user-defined) JavaScript.
 
-### Client-server architecture
+Our experience shows that this hybrid approach gives users the best of both worlds. Early versions of TestCafe ran entirely in the browser. This caused a number of issues. The tests’ capabilities were limited, the application often interfered with page execution, and the sub-optimal logic separation caused browser crashes. Since then, we decoupled the engine, exposed the underlying Node.js API, and implemented a test runner that executes test scripts server-side.
```
[the remark this answers is not in the document]
Comment (titerman) [reply docs/articles/documentation/how-it-works/README.md:18]: I think the original phrasing better serves the promotional purpose of the article tbh

**Comment by titerman** [comment]
@testcafe-build-bot test

**Pull request closed without merging**